### PR TITLE
Networking improvements

### LINKED ISSUE #30: Update networking setup
Given what we've learnt in the last couple of projects, it's time to bring this knowledge back here. 

Let's update the template setup to cover the following topics:
* Allow us to to specify the type that comes back in the response (for example: have a `getComic` method that returns a comic, or error)
* Add a Redux layer (using RTK) that caches the data returned (with a timestamp of the last update) and allows us to have a "refetching" UI (we show the previously cached response while we fetch the latest data)
* Handle timeouts
* Handles auth token refresh with a retry

Considerations:
* As we're using open / free APIs, it might be hard to test/show example of auth token refresh. It might make sense to test it on a closed API and then add to Redbeard with a set of unit tests. We're okay not having a live example of token refresh in the repo, but I'd like to have a (tested) capability of doing so
* I think to handle the "refetching UI" requirement gracefully we'll have to bring back the Response types (Success, Error, etc.)
* I suggest to play with the setup in a separate repo (preferably created from Redbeard template) and once we're all happy with the result - port it back to Redbeard

### PR DESCRIPTION
**Pull request by blazyliszek**
### Description

This partially resolves #30 . From the requirements list this pr is missing caching. Outside requirements, but still related it's missing RemoteData improvements.
I also resolved some of issues that I discovered on the go and  blocked me from moving forward.

### Changes

- replaced sagas with typed sagas / typed backend responses
- fixed found post rn upgrade leftovers
- fixed / improved android build
- fixed test setup
- added persistence for redux (mmkv + encrypted storage for sensitive data)
- added authentication logic
- moved some code around

### Considerations

- I introduced one lint error, that seem to be broken (makeRequest function line 66)
- I just realised it would be nice to mention added flipper plugins in readme
- not that many tests ⏱️
- setApiAuthConfig saga seem to me not so bulletproof. An observer of particular redux field would work better here, but this is not the saga's way? the idea originates from another place where I placed such code in separate mini-middleware  - it was detatched from sagas if someone prefers not to use them.
- `RehydrateAction` type is not 100% accurate, it assumes that we persist reducers only at the top level
- tests prints some console.logs, I left it on purpose - we're gonna handle it once we add a proper Logger.

### Demo

https://github.com/brains-and-beards/react-native-template-redbeard/assets/23454128/a2277c36-0c91-45b0-8c8d-abf719650407

### CODE REVIEW

File: template/.node-version [hunk 1]
```diff
@@ -1 +1 @@
-16
+18
```
Comment (szymonkoper) [inline template/.node-version:1]: @blazlew does it work for you? I use `nvm` and it doesn't use this file. It sets `node` to v20 (my default) in my case. What do you think about using `.nvmrc` instead (or also)?

File: README.md [hunk 2]
```diff
@@ -10,11 +10,20 @@ npx react-native init MyProject --template https://github.com/brains-and-beards/
...
+- `src/screens/LoginScreen.tsx`
 - `src/screens/DemoScreen.tsx`
 - `src/screens/DemoScreen.test.tsx`
 - `src/screens/TranslationsDemoScreen`
 - `src/screens/demoSlice.ts`
+- `src/screens/__tests__/demoSlice.test.ts`
+
+## SAFETY DISCLAIMER
+
+Project contains basic code for auth token management. It stores sensitive information with [react-native-encrypted-storage](https://github.com/emeraldsanto/react-native-encrypted-storage#readme) library. Use at your own risk or replace it with your own implementation.
```
Comment (szymonkoper) [inline README.md]: What is the risk here? Is this disclaimer really needed?

File: template/src/api/common.ts [hunk 3]
```diff
@@ -1,21 +1,166 @@
...
+
+    const hasBody = !!await response.clone().text()
+    const body = hasBody ? await response.json() : null
+
+    if (!response.ok) {
+      const readableBody = body ? ` and ${JSON.stringify(body, null, 2)}` : '';
+      throw new HttpError(`Server returned ${response.status}${readableBody}`, response.status, body);
+    }
+
+    return <T>body;
```
Comment (szymonkoper) [inline template/src/api/common.ts]: It's related to what you wrote in _Considerations_:
> I introduced one lint error, that seem to be broken (makeRequest function line 66)

This `<T>` somehow breaks linter in this file.

With `<T>` linter shows error `Parsing error: Unexpected token eslint` in line 66.
When you remove it then it shows more errors in this file.

File: template/src/redux/persistence.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import EncryptedStorage from 'react-native-encrypted-storage'
+import { MMKV } from 'react-native-mmkv'
+import { initializeMMKVFlipper } from 'react-native-mmkv-flipper-plugin'
+import { Storage } from 'redux-persist'
+import { isJest } from '@utils/env'
+
+const storage = new MMKV()
+
+if (__DEV__ && !isJest) {
+  initializeMMKVFlipper({ default: storage })
```
Comment (szymonkoper) [inline template/src/redux/persistence.ts:9]: I propose to remove `isJest` usage here and globally mock `initializeMMKVFlipper` in tests (for example in `.jest.setup.ts`.

File: template/src/redux/store.ts [hunk 5]
```diff
@@ -1,24 +1,38 @@
...
+import { isJest } from '@utils/env'
 import rootReducer from './rootReducer'
 import rootSaga from './rootSaga'
 
 const sagaMiddleware = createSagaMiddleware()
 
 const middlewares = [sagaMiddleware]
 
-if (__DEV__) {
+if (__DEV__ && !isJest) {
```
Comment (szymonkoper) [inline template/src/redux/store.ts]: Is it possible to remove `isJest` here? Maybe by mocking something

File: template/src/utils/env.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+export const isJest = !!process.env.JEST_WORKER_ID
```
Comment (szymonkoper) [inline template/src/utils/env.ts]: Please try to remove `isJest`. Everywhere where it's needed it can be resolved by mocks.

File: README.md [hunk 7]
```diff
@@ -10,11 +10,20 @@ npx react-native init MyProject --template https://github.com/brains-and-beards/
 
 Project inits with some demo usage files. You can browse them to see the proposed usage patterns. After init feel free to remove them:
 
+- `src/__mocks__/fixtures`
```
Comment (szymonkoper) [inline README.md:13]: This list of demo files looks redundant. Similar content is in `template/README.md`.

What do you think about removing it here? This topic can be moved to a separate issue/PR.

File: README.md (hunk 2 above)
Comment (szymonkoper) [reply README.md]: If it has to stay - is it the right place for this information? Or `template/README.md` is better? There is already `Authentication`.

File: template/src/api/auth.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import { AuthTokens } from './common'
+import { Credentials } from './types/auth.types'
+
+// This is the place where you want to call your auth provider API
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (szymonkoper) [inline template/src/api/auth.ts]: You can remove this eslint and rename:

 - `credentials` to `_credentials`,
 - `refreshToken` to `_refreshToken`.
 
Now if somebody uses this template and wants to modify `logIn` and `refreshTokens` they can miss `eslint-disable` line. I think it will be better with underscore prefixes.

File: template/src/screens/DemoScreen.tsx [hunk 9]
```diff
@@ -26,14 +29,31 @@ interface DemoScreenProps {
...
+
+  const logOut = async () => {
+    try {
+      setIsLogoutLoading(true)
+      // typical logout for apps that require user to be logged in
+      // before giving any further access
+      persistor.pause()
+      await clearPersistence()
+      dispatch(resetStore())
+      persistor.persist()
```
Comment (szymonkoper) [inline template/src/screens/DemoScreen.tsx:52]: How does it work?

```
persistor.pause()
await clearPersistence()
dispatch(resetStore())
persistor.persist()
```

File: template/src/api/types/comic.types.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
+export interface ComicBE {
```
Comment (szymonkoper) [inline template/src/api/types/comic.types.ts:1]: Maybe it's a good idea to use `*.d.ts` extension instead of `*.types.ts` for files consisting of types declarations only (no implementation codes).

If you agree then it can be a separate Issue/PR.

File: template/src/config/ui/typography.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import { StyleSheet, TextStyle } from 'react-native'
```
Comment (szymonkoper) [inline template/src/config/ui/typography.ts:1]: How is typography related to "Networking improvements"? Why is it in this PR? I think `typography.ts` and `DemoTextInput.tsx` could be in a separate PR.

I miss some information about what is this file and how to use it. Maybe it makes sense to create another Issue or PR for describing it?

File: template/src/redux/store.ts [hunk 12]
```diff
@@ -1,24 +1,38 @@
...
   middlewares.push(createDebugger())
 }
 
 const store = configureStore({
   reducer: rootReducer,
-  middleware: getDefaultMiddleware => getDefaultMiddleware({ thunk: false }).concat(middlewares),
+  middleware: getDefaultMiddleware =>
+    getDefaultMiddleware({
+      serializableCheck: {
+        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
```
Comment (szymonkoper) [inline template/src/redux/store.ts:23]: I didn't know why those actions are ignored. I googled it and found it in [Redux Toolkit docs](https://redux-toolkit.js.org/usage/usage-guide/#use-with-redux-persist), but it's not easy to understand just by quickly looking at the code.

Can you extract `[FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER]` and name it for example `reduxPersistActions`?
So that it gives a hint what are those actions and why are they ignored here.

File: template/src/redux/rootReducer.ts [hunk 13]
```diff
@@ -1,7 +1,25 @@
...
+  return appReducer(state, action)
 }
 
-export default rootReducer
+const rootPersistConfig = {
+  key: 'root',
+  storage,
+  blacklist: ['auth'],
+}
+export default persistReducer(rootPersistConfig, rootReducer as typeof appReducer)
```
Comment (szymonkoper) [inline template/src/redux/rootReducer.ts]: Please remove `as typeof appReducer` here.

`rootReducer` can be defined like this:

```
const rootReducer: typeof appReducer = (state, action) => {
```

File: template/src/redux/rootReducer.ts [hunk 14]
```diff
@@ -1,7 +1,25 @@
...
+import { storage } from './persistence'
+import { resetStore } from './rootActions'
 
-const rootReducer = {
+const appReducer = combineReducers({
+  auth: authReducer,
   demo: demoReducer,
+})
+
+const rootReducer = (state: ReturnType<typeof appReducer>, action: AnyAction) => {
```
Comment (szymonkoper) [inline template/src/redux/rootReducer.ts]: By just looking at names: `appReducer` vs `rootReducer` I don't know what is the difference.

What do you think about renaming it to something like `resettableAppReducer `, `appReducerWithReset` or any other name that will indicate how is it different from `appReducer`.

File: template/src/redux/rootReducer.ts [hunk 15]
```diff
@@ -1,7 +1,25 @@
...
+    return appReducer(undefined, action)
+  }
+  return appReducer(state, action)
 }
 
-export default rootReducer
+const rootPersistConfig = {
+  key: 'root',
+  storage,
+  blacklist: ['auth'],
```
Comment (szymonkoper) [inline template/src/redux/rootReducer.ts:23]: By default we don't want to persist auth tokens? Will users have to sign in again each time app is restarted or do I miss something? Is it what we want?

File: template/src/api/authSlice.ts [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+import { REHYDRATE, persistReducer } from 'redux-persist'
+import { PersistPartial } from 'redux-persist/es/persistReducer'
+import { call, put, takeLatest, takeLeading } from 'typed-redux-saga'
+import { Failure, Loading, NotRequested, RemoteData, Success, isSuccess } from '@api/RemoteData'
+import { logIn as logInRequest } from '@api/auth'
+import { AuthTokens, setAuthConfig } from '@api/common'
+import { Credentials } from '@api/types/auth.types'
+import { safeStorage } from '@redux/persistence'
+import { resetStore } from '@redux/rootActions'
+import { RootState } from '@redux/store'
```
Comment (szymonkoper) [inline template/src/api/authSlice.ts:11]: I think it should not be importing `RootState`, because it causes requirement cycles: `authSlice.ts` imports `@redux/store`, which then again imports this `authSlice.ts`.

I propose to define local `RootState` type here, which will contain only what's used by this `authSlice.ts` and nothing more.

BTW this project is not checking for requirement cycles. Maybe it makes sense to add an issue and take a look at it later in another PR. What do you think?

File: template/src/api/common.ts [hunk 17]
```diff
@@ -1,21 +1,166 @@
...
+
+export function setAuthConfig(newConfig: AuthConfig) {
+  Object.assign(authConfig, newConfig)
+};
+
+
+async function makeRequest<T>(params: RequestParams, isFirstTry = true): Promise<T> {
+  // Change the default request timeout
+  const controller = new AbortController();
+  const timeoutId = setTimeout(() => controller.abort(), API_TIMEOUT);
```
Comment (szymonkoper) [inline template/src/api/common.ts]: 👏

**Review by szymonkoper — CHANGES_REQUESTED** [review]
Generally, I like the direction of networking changes. It looks better than before. 👍  I left a few comments

Please next time try to do smaller PRs. 🥺 What really is in scope of “Networking improvements” could be as well split into few smaller PRs to be reviewed and merged in steps. Additionally, there are some things that are not related to “Networking improvements” that could be done in a separate PR.

File: template/.node-version (hunk 1 above)
Comment (blazyliszek) [reply template/.node-version:1]: moved to nozbe

File: README.md (hunk 2 above)
Comment (blazyliszek) [reply README.md]: I reduced harshness of the message, the risk is that it's a community-maintained package.  For some cases you want to have full control over security features. I'll keep it in both readmes with a little bit different info.

File: template/src/api/common.ts (hunk 3 above)
Comment (blazyliszek) [reply template/src/api/common.ts]: Thanks, I'll use `as T` there instead.

File: template/src/redux/persistence.ts (hunk 4 above)
Comment (blazyliszek) [reply template/src/redux/persistence.ts:9]: you're right

File: README.md (hunk 7 above)
Comment (blazyliszek) [reply README.md:13]: one readme you see once browsing template's repo, the other one gets copied into your project. It might be redundant and not, I didn't split it this way so I'm leaving both for now

File: template/src/api/auth.ts (hunk 8 above)
Comment (blazyliszek) [reply template/src/api/auth.ts]: hmm, ok replacing thanks

File: template/src/screens/DemoScreen.tsx (hunk 9 above)
Comment (blazyliszek) [reply template/src/screens/DemoScreen.tsx:52]: it pauses redux-persist
it clears the persistence
it resets the store to initial state
it resumes redux-persist

a nicer way would be to use built-in redux-persist action `PURGE`, but it happened to be buggy

File: template/src/api/types/comic.types.ts (hunk 10 above)
Comment (blazyliszek) [reply template/src/api/types/comic.types.ts:1]: I think `d.ts` files were meant to provide types to incorrectly typed / untyped JS libraries.
[This is](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html) what official examples looks like.
Buta at the same time I'm not 100% sure about that, or if there is another con of using it differently.
leaving for now

File: template/src/config/ui/typography.ts (hunk 11 above)
Comment (blazyliszek) [reply template/src/config/ui/typography.ts:1]: it happend to be here, cause I needed some textInputs to showcase the login. I'll add a comment to this file. it was meant to hold app text styles.

File: template/src/redux/store.ts (hunk 12 above)
Comment (blazyliszek) [reply template/src/redux/store.ts:23]: it's fine, you can tell what is it just by looking at the imports.
we can change it if there are more actions to ignore  from different sources.

File: template/src/redux/rootReducer.ts (hunk 13 above)
Comment (blazyliszek) [reply template/src/redux/rootReducer.ts]: Not exactly, if I change it like this it's giving me the following TS error on changed line:
```
Type '(state: ReturnType<typeof appReducer>, action: AnyAction) => CombinedState<{ auth: AuthState & PersistPartial; demo: DemoState; }>' is not assignable to type 'Reducer<CombinedState<{ auth: AuthState & PersistPartial; demo: DemoState; }>, AnyAction>'.
  Types of parameters 'state' and 'state' are incompatible.
    Type 'CombinedState<{ auth: AuthState & PersistPartial; demo: DemoState; }> | undefined' is not assignable to type 'CombinedState<{ auth: AuthState & PersistPartial; demo: DemoState; }>'.
      Type 'undefined' is not assignable to type 'CombinedState<{ auth: AuthState & PersistPartial; demo: DemoState; }>'.
        Type 'undefined' is not assignable to type 'EmptyObject'.ts(2322)
```

File: template/src/redux/rootReducer.ts (hunk 14 above)
Comment (blazyliszek) [reply template/src/redux/rootReducer.ts]: hmm, I see with the typescript fixes it got a little bit too confusing.
I'm changing it to `rootReducer` and `resettableRootReducer`

Comment (blazyliszek) [reply template/src/redux/rootReducer.ts]: or I'll change it rather to `rootReducerWithReset` , cause reducer is just a function and this one resets store to initial sate.

File: template/src/redux/rootReducer.ts (hunk 15 above)
Comment (blazyliszek) [reply template/src/redux/rootReducer.ts:23]: AuthSlice have it's own persist config and it's persisted separately in encrypted storage. If we don't blacklist it here, it's gonna be stored in 2 places at the same time

File: template/src/api/authSlice.ts (hunk 16 above)
Comment (blazyliszek) [reply template/src/api/authSlice.ts:11]: It's not an issue for typescript types. https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export

Comment (blazyliszek) [reply template/src/api/authSlice.ts:11]: we have similar type circular dependency for react navigation screen param types.

**Comment by blazyliszek** [comment]
> Generally, I like the direction of networking changes. It looks better than before. 👍 I left a few comments
> 
> Please next time try to do smaller PRs. 🥺 What really is in scope of “Networking improvements” could be as well split into few smaller PRs to be reviewed and merged in steps. Additionally, there are some things that are not related to “Networking improvements” that could be done in a separate PR.

Yeah, sorry for that, It was partially based on another thing which I had to prove in a broader context. Aaand it got a little bit out of hand.

File: template/src/redux/rootReducer.ts (hunk 13 above)
Comment (blazyliszek) [reply template/src/redux/rootReducer.ts]: allright, I missed the part with no explictly typing the arguments, now it works! thanks!

File: template/src/api/authSlice.ts (hunk 16 above)
Comment (szymonkoper) [reply template/src/api/authSlice.ts:11]: I have created an issue for this -> https://github.com/brains-and-beards/react-native-template-redbeard/issues/34

**Review by szymonkoper — APPROVED** [review]
Looks good 👍  All issues resolved or extracted to a separate topics

**Comment by blazlew** [comment]
add auth token setup; improve on networking types; fix build / test issues
